### PR TITLE
[Model] Support languages added by fine-tuned Qwen3-TTS checkpoints

### LINKED ISSUE #4170: [Feature]: Allow Custom Languages in `serving_speech.py` to Support Fine-Tuned Qwen3-TTS Models
### 🚀 The feature, motivation and pitch

At present, we maintain a hardcoded list of supported languages for Qwen3-TTS. However, this may not work correctly when running inference with fine-tuned Qwen3-TTS models that support languages beyond those currently specified.

Since the Qwen3-TTS checkpoint `config.json` already defines the supported languages through `codec_language_id`, would it be possible to infer the supported languages from `config.json` during validation? This would help ensure compatibility with fine-tuned models that support additional languages.

Ref: https://github.com/vllm-project/vllm-omni/blob/6cefe913703fa67a0726c327476334fe391154d0/vllm_omni/entrypoints/openai/serving_speech.py#L102-L114

https://huggingface.co/Qwen/Qwen3-TTS-12Hz-0.6B-Base/blob/main/config.json#L115-L125

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by n0n4m39911**
## Purpose

Closes #4170 

Language validation for Qwen3-TTS was a hardcoded list (`_TTS_LANGUAGES`) of the 10
official languages + `Auto`. Fine-tuned checkpoints that add languages or dialects
to `talker_config.codec_language_id` (e.g. `beijing_dialect`) were rejected at the API
layer with `Invalid language`, even though the model supports them.

This PR derives the supported set from the checkpoint instead:

- At server init, `_load_supported_languages()` reads
`talker_config.codec_language_id`
  from the model config (handles both object and plain-dict `talker_config`),
  title-cases the keys, and always adds `Auto`.
- Request validation is now case-insensitive: the request's `language` is
normalized to
  the title-cased config form (`english` / `ENGLISH` / `English` are equivalent)
before
  matching, so the normalized value is what flows downstream.
- Fallbacks: if `codec_language_id` is missing/empty or the model config can't be
read,
  validation falls back to the previous hardcoded default list (with a warning).
  Non-`qwen3_tts` model types also keep the default list, preserving existing
behavior
  for models that share the fallback validator.

Docs (`docs/serving/speech_api.md`) updated to describe the config-driven behavior.

## Test Plan

Unit tests added in `tests/entrypoints/openai_api/test_serving_speech.py`
(`TestTTSMethods`) covering:

- Loading languages from `codec_language_id` (object and plain-dict `talker_config`),
  including a custom dialect, title-casing, and the implicit `Auto`
- Case-insensitive validation + normalization of `request.language`
- Rejection of languages not in the configured set
- Fallback to the default list: empty `codec_language_id`, unreadable model
  config, and non-`qwen3_tts` model types

## Test Result

pytest tests/entrypoints/openai_api/test_serving_speech.py

```
151 passed, 0 failed
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [x] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by n0n4m39911** [comment]
@linyueqian , Could you please review this PR.

**Review by hsliuustc0106 — COMMENTED** [review]
This PR cleanly addresses the limitation of hardcoded language validation for Qwen3-TTS. The configuration-driven approach with fallback behavior is well-considered, and the test coverage is comprehensive.

**Non-blocking suggestion:**

The validation method mutates `request.language` in-place during normalization. This works correctly here (verified by the tests), but mutating request params during validation can be surprising to readers. Consider either:
- Normalizing early in the request handling pipeline before validation
- Or returning the normalized value from the validation method and assigning it at the call site

This is a minor style point and not a blocker — the current behavior is correct and well-tested.

**Review by linyueqian — COMMENTED** [review]
Thanks for picking this up @n0n4m39911, the implementation matches what we discussed in #4170 and the test coverage is thorough.

### What works well

- `_load_supported_languages` reads from the same `talker_config.codec_language_id` that Stage 0 already consults at `prompt_embeds_builder.py:962` and again at L1464, so the API surface and the model now agree on the allow-list (which is the whole point).
- `.title()` normalization is safe: Stage 0 does `codec_language_id.get(language.lower())`, so any casing flows through correctly downstream. The `"beijing_dialect" → "Beijing_Dialect"` round-trip is the right invariant to lock in, and the test does so explicitly.
- The fallback covers the three meaningful cases (empty `codec_language_id`, unreadable `model_config`, non-`qwen3_tts` types), so other types that fall through to `_validate_qwen_tts_request` (e.g. `higgs_audio_v3`) keep their existing behavior.

### On the mutation comment

Re. @hsliuustc0106's note about in-place mutation of `request.language`: I'd keep the current pattern. `_validate_voxcpm2_request` already does the same thing (`request.voice = request.voice.lower()` at `serving_speech.py:1301`), so changing only this call site would diverge from the surrounding convention. If we want return-and-assign across the board, a dedicated follow-up seems like the right scope.

### Nits (non-blocking)

🟢 `_TTS_LANGUAGES` is a module-level mutable `set`, and three of the four fallback paths return it directly, so `self.supported_languages is _TTS_LANGUAGES`. Nothing mutates it today, but making `_TTS_LANGUAGES` a `frozenset` (or copying on assignment) would be safer against future writers.

💡 `isinstance(codec_language_id, dict)` rejects non-`dict` mappings. HF `PretrainedConfig` only emits plain dicts so this is fine in practice, but `isinstance(codec_language_id, Mapping)` from `collections.abc` would be a touch more general if you want to future-proof.

### Before merge

Please run `uvx ruff check --fix` + `uvx ruff format` on the touched files. The project's pre-commit gates style and a few recent PRs have bounced on it.

LGTM otherwise, happy to approve once style is clean.

**Comment by n0n4m39911** [comment]
Thanks for the review @linyueqian , @hsliuustc0106 

As suggested, I updated `_TTS_LANGUAGES` to use a frozen set and also ran Ruff checks and formatting on the modified files. Please take a look at the latest changes when you have a chance.

@hsliuustc0106, as @linyueqian suggested, I've decided to keep the in-place mutation of `request.language` for now.

**Review by linyueqian — APPROVED** [review]
lgtm

**Comment by n0n4m39911** [comment]
@linyueqian , Could you please merge this PR
